Handle config file subdirectories properly

phillipjohnston opened the pull request:
@legeana,

With the adjustment to a loop, the variables aren't updated as they were before due to Meson having immutable variables. We need to call subdir() in the loop, and use the generic config variable for that run. I didn't realize this until I did some heavy testing across systems - OS X built fine, but Linux was suddenly failing to find headers like `signal.h`, even though they were properly detected.

With this change, I see the configuration headers populated correctly once again.

phillipjohnston committed:
Handle config file subdirectories properly

With the adjustment to a loop, the variables aren't updated as they were before due to Meson having immutable variables. We need to call subdir() in the loop, and use the generic config variable for that run

nirbheek commented:
lgtm

legeana commented on meson.build:
```diff
@@ -39,13 +39,15 @@ cc_dict = {
     'compiler': meson.get_compiler('c'),
     'machine': host_machine,
     'config': conf,
```
`conf` and `conf_native` are no longer used. Please delete them from the lines 35 and 36.
Also it appears that the `config` key in the dictionary is not needed and you can just create an empty `configuration_data()` object in the loop instead of getting an empty version from the dict.

What do you think?

phillipjohnston replied:
Roger

legeana commented on meson.build:
```diff
@@ -39,13 +39,15 @@ cc_dict = {
     'compiler': meson.get_compiler('c'),
     'machine': host_machine,
     'config': conf,
+    'header-subdir': 'private',
```
Would you mind renaming it into `config_h_subdir`, `buildtype_subdir` or equivalent? Header is a bit too vague.

phillipjohnston replied:
Roger

legeana commented on meson.build:
```diff
@@ -181,14 +183,15 @@ endif
   endif
 
+  # Execute subdir to create config.h for this pass
+  subdir(entry.get('header-subdir'))
```
Could you please add a comment that subdir() relies on the `config` variable being set in the loop? It's not quite obvious otherwise and a future maintainer might miss the context.

phillipjohnston replied:
Roger.

phillipjohnston committed:
Refactor configuration data update

- Rename header variable in dictionary for clearer purpose
- Eliminate unnecessary conf and conf_native variables by moving configuration_data() into the loop
- Add clarifying comments to subdir calls

phillipjohnston commented:
@legeana, changes made as requested. Tested locally with a project using CMocka as a subproject, configuration header output still generated properly.
